### ISSUE TITLE
Support Parquet data format

### ISSUE THREAD

**Issue by roll** [frictionlessdata/frictionless-py#477]
# Overview

Parquet integration is a really popular feature request for Frictionless. We want to have integration. At the same time, it's not been discovered yet so this issue requires a design solution proposal. One idea that it can be implemented using `pandas` - https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html

Usually, integration means that we can:

> for resource (internally using `Parser`)

```
# Read
resource = Resource('parquet-file')
resource.read_rows()
# etc

# Write
resource = Resource('table.csv')
resource.write('parquet-file')
```

# Plan
- [ ] research what is Parquet format and how it can be mapped to Frictionless primitives (package/resource/schema) ping @roll to sync
- [ ] TBD

**Comment by lwinfree** [comment]
+1 we have users ask us about this a few times a year, so this would be great

**Comment by zaneselvans** [comment]
I think we'd make use of this in @catalyst-cooperative / PUDL for publishing our long tables.

**Comment by rufuspollock** [comment]
@roll think this is an interesting one. Can you say a bit more about what "supporting" would mean (that would also help for same question for other formats).

**Comment by roll** [comment]
Hi @rufuspollock 

For a platform like Zenodo "support" would mean an ability to have:
```
Resource/Package.from_zenodo(...)
resource/package.to_zenodo(...)
```
And for a file format, an ability to read/write:
```
resource = Resource(path='table.parquet')  # not sure if it's applicable to Parquet file format but guess so
resource.write(path='table.parquet')
```
Also, supporting Dropbox/Google Drive/etc can be just a scheme support as we have for S3:
```
resource = Resource(path='dropbox://table.csv')
resource.write(path='dropbox://table.csv')
```

**Comment by zaneselvans** [comment]
For integration with Zenodo (#484), in addition to being able to read/write packages from/to a deposition, it would be very useful to be able to set the deposition level metadata appropriately. Some of the fields overlap with those that you might find in a data package, but [the depositions also have their own elements](https://developers.zenodo.org/#entities) which can be set through the Zenodo API.

Of course Parquet files can already be used as resources in a generic datapackage, so I was imagining that "support" would mean allowing them to be used as tabular resources in a tabular datapackage, since they represent tabular data with their own column-wise schema and metadata. Though this might mean duplicating the metadata in both the parquet files and the datapackage / resource descriptors. But I don't think Parquet currently allows relationships / constraints between tables as could be described with a TableSchema.

**Comment by khughitt** [comment]
The API for interacting with parquet/[feather](https://ursalabs.org/blog/2020-feather-v2/) files in Python and Pandas is quite similar, so you may be able to include support for both, with little extra effort, which would be quite nice!

**Comment by antoniocali** [comment]
+1 :) It would be a great plugin feature

**Comment by adriangb** [comment]
Any updates @roll ? Parquet is a pretty popular tabular format. I suspect you can't support _all_ parquet features, but it shouldn't be too hard to support basic usage.

**Comment by roll** [comment]
We'd like to implement this and a few others most popular feature requests in 2021 cc @lwinfree 

Also, if someone is interested in working on it please let us know :+1:

**Comment by kindly** [comment]
@roll 
I am happy to help here. Some work has been done on converting a CSV datapackage to PARQUET.  https://github.com/kindly/datapackage_convert_py/

I have not found any other python libraries that currently stream CSV data into PARQUET format in a memory efficient way.
There are execution engines/dbs that will do this e.g `pyspark`, `dask` and `duckdb` but I think these are too heavyweight to include in the framework.

A `describe` for parquet files would be possible but not sure how useful it is considering the schema is also held within the files.
`parquet` to other formats like CSV, SQL should be possible but I am not sure of any libarary that could do this in streaming way apart from the ones above.

**Comment by roll** [comment]
Hi @kindly,

It would be great. I'm going to review this issue on Monday and update the implementation hints (regarding Frictionless internals)

**Comment by roll** [comment]
Hi @kindly @shashigharti @aivuk

I started analyzing this issue and ended up with the fact that it seems we can add support of it as easy as here - https://github.com/frictionlessdata/frictionless-py/pull/1186

Can you please try it?

```
git clone git@github.com:frictionlessdata/frictionless-py.git
git checkout origin/477/support-parquet
make install
frictionless extract data/table.parq  # for example
# do you have other files to test?
```

**Comment by shashigharti** [comment]
nice! it works. I used both data 'table.parq' and '[userdata1.parquet](https://github.com/Teradata/kylo/blob/master/samples/sample-data/parquet/userdata1.parquet)'
![parquet](https://user-images.githubusercontent.com/5582809/179709526-af5a26a1-fd5b-4d90-82d0-296490f68249.png)

**Comment by roll** [comment]
It's been implemented in v5 (#1186) (will be released this month)

I also created a follow-up issue - https://github.com/frictionlessdata/frictionless-py/issues/1203
